src/Razor/benchmarks/Microsoft.AspNetCore.Razor.Microbenchmarks/LanguageServer/RazorDiagnosticsBenchmark.cs

```diff
@@ -0,0 +1,227 @@
+        }
+    }
+
+    protected internal override void Builder(IServiceCollection collection)
+    {
+        collection.AddSingleton<ClientNotifierServiceBase, NoopClientNotifierServiceForDiagnostics>();
+        collection.AddSingleton<IOnInitialized, NoopClientNotifierServiceForDiagnostics>();
+    }
+
+    private protected override LanguageServerFeatureOptions BuildFeatureOptions()
+    {
+        return new LanguageServerFeatureOptionsForDiagnostics();
+    }
+
+    private string GetFileContents(FileTypes fileType)
+    {
+        var sb = new StringBuilder();
+
+        sb.Append("""
+            @using System;
+            """);
+
+        for (var i = 0; i < (fileType == FileTypes.Small ? 1 : 100); i++)
+        {
+            sb.Append($$"""
+            @{
+                var y{{i}} = 456;
+            }
+
+            <div>
+                <p>Hello there Mr {{i}}</p>
+            </div>
+            """);
+        }
+
+        sb.Append("""
+
+             <div></div>
+
+             @functions
+             {
+                public void M()
+                {
+                    CallOnMe();
+                }
+             }
+
+             """);
+
+        return sb.ToString();
+    }
+
+    [Benchmark(Description = "Diagnostics")]
+    public async Task RazorDiagnosticsAsync()
+    {
+        var request = new VSInternalDocumentDiagnosticsParams
+        {
+            TextDocument = new TextDocumentIdentifier
+            {
+                Uri = DocumentUri!
+            },
+        };
+
+        await DocumentPullDiagnosticsEndpoint!.HandleRequestAsync(request, RazorRequestContext, CancellationToken.None);
+    }
+
+    [GlobalCleanup]
+    public async Task CleanupAsync()
+    {
+        File.Delete(_filePath!);
+
+        var innerServer = RazorLanguageServer.GetInnerLanguageServerForTesting();
+
+        await innerServer.ShutdownAsync();
+        await innerServer.ExitAsync();
+    }
+
+    private class LanguageServerFeatureOptionsForDiagnostics : LanguageServerFeatureOptions
+    {
+        public override bool SupportsFileManipulation => true;
+
+        public override string ProjectConfigurationFileName => "LanguageServerConstants.DefaultProjectConfigurationFile";
+
+        public override string CSharpVirtualDocumentSuffix => ".ide.g.cs";
+
+        public override string HtmlVirtualDocumentSuffix => "__virtual.html";
+
+        public override bool SingleServerCompletionSupport => false;
+
+        public override bool SingleServerSupport => true;
+
+        public override bool SupportsDelegatedCodeActions => true;
+
+        // Code action and rename paths in Windows VS Code need to be prefixed with '/':
+        // https://github.com/dotnet/razor/issues/8131
+        public override bool ReturnCodeActionAndRenamePathsWithPrefixedSlash
+            => RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+    }
+
+    private class NoopClientNotifierServiceForDiagnostics : ClientNotifierServiceBase
```

Comment:
Nit: This is not a "no op" notifier any more 😛

Since these are nested types, I think its fine to just call this `ClientNotifierService` and the above `FeatureOptions` or something, the "ForDiagnostics" is redundant